CI: Speed up static checks by checking only changed files

akien-mga opened the pull request:
Our static checks used to take close to 8m to run, as many of the checks run on most files in the codebase each time, and are relatively slow bash or python scripts working file by file.

This PR optimizes the workflow by changing the scripts to (optionally) operate on a list of files, and we pass in the list of files actually changed in the PR.

We disable the dotnet format check if no `modules/mono` files are changed, as it's very slow and couldn't be optimized the same way as others. And we do the same for the JS for an extra 15 s speedup.

What's left as potential to optimize would be the dependencies install. Maybe the pip stuff can be cached too (see [actions/setup-python](https://github.com/actions/setup-python#caching-packages-dependencies)).

Before:
![image](https://user-images.githubusercontent.com/4701338/233152362-c87af42b-f0e7-40c3-aae0-1d203feabfef.png)

After:
![image](https://user-images.githubusercontent.com/4701338/233154638-14a18699-ea3b-4e0a-bdf0-bb1ca8efb3a2.png)

So we go from 7m 45s to ~1m 5s~ **50s**.

akien-mga committed:
CI: Speed up static checks by checking only changed files

- file_format, header_guards and clang-format benefit from this short list.
- dotnet-format, Python and JS checks don't, but they're only relevant for
  PRs changing a specific set of files, so we skip them when those files
  aren't modified.

The logic to get changed files only works reliably for:
- Pull request events
- Non-force pushed push events

So when force pushing a branch in your fork, or creating a new branch,
it will still scan all files as fallback.

Upgraded CI runner to Ubuntu 22.04 so we get clang-format 14 out of the box,
so we don't need to install a custom version (saves ~15 s). We also cache
the APT dependencies to speed up the build and avoid flaky Ubuntu/Microsoft
repos.

Calinou commented:
Another possible optimization is to run each static check in parallel, and abort as soon as one of them reports an error.

aaronfranke reviewed (approved):
Sounds good to me. Running fast is important now that static checks block the other jobs. And it's great that the list of files is optional, I like being able to independently run the script on all files so this is the best of both worlds.

RedworkDE commented:
Possible edge case: Some huge PR with thousands of changed files could exceed the maximum size of the argument list for the scripts and fail that way. It might be better to pass the file with the changed files instead of the changed files directly.

akien-mga commented:
> Possible edge case: Some huge PR with thousands of changed files could exceed the maximum size of the argument list for the scripts and fail that way. It might be better to pass the file with the changed files instead of the changed files directly.

Updated to do this.

> What's left as potential to optimize would be the dependencies install. Maybe the pip stuff can be cached too (see [actions/setup-python](https://github.com/actions/setup-python#caching-packages-dependencies)).

I reduced the setup time for dependencies by moving the static checks to Ubuntu 22.04, which provides clang-format 14 which is recent enough, so we don't need to install it from the LLVM repo anymore.

It's down from ca. 30 s to 15 s in average. Though sometimes it still takes longer if for whatever reason the Ubuntu or pip repos are slow. We could attempt to cache pip packages, but we should see if the `setup-python` action's own setup wouldn't be slower than the time we can save with caching.

The codespell action also sets up an older version of Python internally, but I'm not sure we can control that and share the python versions.

akien-mga commented:
> I reduced the setup time for dependencies by moving the static checks to Ubuntu 22.04, which provides clang-format 14 which is recent enough, so we don't need to install it from the LLVM repo anymore.
> 
> It's down from ca. 30 s to 15 s in average. Though sometimes it still takes longer if for whatever reason the Ubuntu or pip repos are slow, like this: ![image](https://user-images.githubusercontent.com/4701338/233312133-d0df7431-7852-4990-be40-7aaf10229a9a.png)

To reduce this further, I'm adding https://github.com/marketplace/actions/cache-apt-packages to cache the APT packages. This seems to reduce it more consistently to around 9-12 s for the current packages we have setup, which is a bit faster than a "good" APT run, but also prevents us from getting into issues with either Ubuntu or Microsoft repositories being wonky.

> We could attempt to cache pip packages, but we should see if the `setup-python` action's own setup wouldn't be slower than the time we can save with caching.

Tried that, it works but it's indeed slower.

Installing our current pip dependencies straight with the pip from the Ubuntu 22.04 image takes ~6-7 s, while setting up python again with pip cache + restoring dependencies from cache seems to take between 12 and 18 s in my tests.

---

Final result, I think I'm stopping here :)

**54s for a trivial build.**

Should scale well with more files touched in bigger PRs, linearly to a max a bit below what it used to be (maybe 7m if changing all files in the codebase).

![image](https://user-images.githubusercontent.com/4701338/233323998-212c72e7-dc2f-478c-b4c0-b9778c988ebb.png)

akien-mga commented:
Tweaked the job name as the previous list of scripts was incomplete, so I simplified it:
![image](https://user-images.githubusercontent.com/4701338/233327592-12ee2903-3b86-4c3c-9267-4a7ba9e592bb.png)

And also added conditions to skip Python checks when no Python files are changed, shaves off a few seconds more.

I'm done force push tweaks :)

YuriSizov commented:
Cherry-picked for 4.0.3.
